Wrappers/Python/cil/optimisation/functions/SAGFunction.py

```diff
@@ -0,0 +1,161 @@
+   and the gradient function returns the approximate gradient. This can be used with the `cil.optimisation.algorithms` algorithm GD to give a stochastic optimisation method.  
+   By incorporating a memory of previous gradient values the SAG method can achieve a faster convergence rate than black-box stochastic gradient methods. 
+
+    Parameters:
+    -----------
+    functions : `list`  of functions
+                A list of functions: :code:`[F_{0}, F_{1}, ..., F_{n-1}]`. Each function is assumed to be smooth function with an implemented :func:`~Function.gradient` method. Each function must have the same domain. The number of functions must be strictly greater than 1. 
+    sampler: An instance of one of the :meth:`~optimisation.utilities.sampler` classes which has a `next` function implemented and a `num_indices` property.
+        This sampler is called each time gradient is called and  sets the internal `function_num` passed to the `approximate_gradient` function.  The `num_indices` must match the number of functions provided. Default is `Sampler.random_with_replacement(len(functions))`. 
+    
+    Note
+    ------
+    
+    The user has the option of calling the class method `warm_start_approximate_gradients` after initialising this class. This will compute and store the gradient for each function at an initial point. If this method is not called, the gradients are initialised with zeros. 
+
+    References
+    ----------
+    Schmidt, M., Le Roux, N. and Bach, F., 2017. Minimizing finite sums with the stochastic average gradient. Mathematical Programming, 162, pp.83-112. https://doi.org/10.1007/s10107-016-1030-6
+
+    """
+
+    def __init__(self, functions,  sampler=None):
+        self.list_stored_gradients = None
+        self.full_gradient_at_iterate = None
+        self._warm_start_just_done=False
+
+        super(SAGFunction, self).__init__(functions, sampler)
+
+
+        
+
+    def approximate_gradient(self, x, function_num,  out=None):
+        """ Returns the gradient of the selected function or batch of functions at :code:`x`. 
+            The function_num is selected using the sampler. 
+
+        Parameters
+        ----------
+        x: element in the domain of the `functions`
+
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+
+        """
+        if self.list_stored_gradients is None: 
+            self.list_stored_gradients = [
+                0*x for fi in self.functions]
+            self.full_gradient_at_iterate = 0*x
+            
+
+        self.stoch_grad_at_iterate = self.functions[function_num].gradient(x)
+
+        if self._warm_start_just_done:
+            self._update_data_passes_indices(list(range(self.num_functions))+ [self.function_num])
+            self._warm_start_just_done=False 
+        else:
+            self._update_data_passes_indices([self.function_num])
```

Comment:
This code is relevant only for the first iteration. 

I suggest 
1. to remove all the if statements,
2. to call `_update_data_passes_indices(list(range(self.num_functions)))` in `warm_start_approximate_gradient`
3. modify in this class the property `data_passes_indices` to return what you expect

probably this will do what you want
```python

data_passes_indices = _data_passes_indices[:]
if len(data_passes_indices[0]) == self.num_functions:
    a = data_passes_indices.pop(1)
    data_passes_indices[0] += [a]
```

```

Reply:
Done